impress-org/givewp
# Stripe Webhooks can now be configured from the Setup Page

Resolves #4949 

## Description



- Adds a setup step to configure Stripe Webhooks.
- Provides a deep link to the webhook configuration on the Stripe dashboard.
- Provides a copy-able webhook listener URL.
- Adds an endpoint to check if webhooks are configured.
- Ensures that the Stripe Gateway is enabled so it can listen for webhooks.
- Creates a test customer for triggering test events/webhooks.
- Triggers test `customer.updated` events/webhooks.
- Adds JS polling to confirm webhooks are configured.
- Deactivates "Configure Webhooks" link if already configured.

## Affects



- Ensures that the Stripe Gateway is enabled so it can listen for webhooks. This flow configures the Stripe gateway, but it wasn't automatically enabled.

## Visuals



Connect to Stripe
![image](https://user-images.githubusercontent.com/10858303/89460980-31648380-d739-11ea-995d-0522ae634483.png)

Configure Webhooks
![image](https://user-images.githubusercontent.com/10858303/89460860-feba8b00-d738-11ea-934e-2aea58e18ae2.png)

Copy Webhook URL
![image](https://user-images.githubusercontent.com/10858303/89460881-0bd77a00-d739-11ea-9146-f2b0e1685c4f.png)

Checking webhooks animation
![Peek 2020-08-05 18-30](https://user-images.githubusercontent.com/10858303/89470446-d0917700-d749-11ea-95fb-58140e3ed6c9.gif)


Webhooks configured
![image](https://user-images.githubusercontent.com/10858303/89460919-1eea4a00-d739-11ea-9e9e-785cbbbf75cb.png)


## Pre-review Checklist




## Testing Instructions



1. Navigate to the Setup Page.
1. Connect to Stripe.
1. Copy the webhooks URL.
1. Click "Configure Webhook".
1. Configure the webhook in the Stripe Dashboard.
1. Return to the Setup Page.
1. The "Configure Webhooks" button should change to "Webhooks Configured!" and be disabled.

**Note:** You'll need to establish a tunnel to forward Stripe webhooks for local development. Local by Flywheel has ngrok support, but otherwise free ngrok doesn't support local domain forwarding. I can recommend [Expose by Beyond Code](https://beyondco.de/docs/expose/introduction), an open source tunnel application written in PHP. While Expose can be self hosted, Beyond Code offers a free hosted version with their free account.

## User Documentation
